### DISCUSSION TITLE
not able to get parent span, send from Client service to Server service

### DISCUSSION THREAD

**Discussion by mayur-more** [open-telemetry/opentelemetry-java#3255] [Q&A]
I am new to OpenTelemetry word. I have created spans for my services separately, but when i am try to combine spans of two different services, using context propogation, I am not able to do it successfully. I have used following code:

`// at client side:
  public static void sendContext(String resource) {

TextMapSetter<HttpURLConnection> setter =
    new TextMapSetter<HttpURLConnection>() {
      @Override
      public void set(HttpURLConnection carrier, String key, String value) {
        carrier.setRequestProperty(key, value);
      }
    };
HttpURLConnection transportLayer = null;
String urlString = "http://127.0.0.1:8080" + resource;
try {
  URL url = new URL(urlString);
  transportLayer = (HttpURLConnection) url.openConnection();
} catch (MalformedURLException ex) {
  System.out.println(ex.getMessage());
} catch (IOException e) {
  System.out.println(e.getMessage());
}
GlobalOpenTelemetry.getPropagators()
    .getTextMapPropagator()
    .inject(Context.current(), transportLayer, setter);
}

// at server side:
public static Context getContext(HttpServletRequest request) {

TextMapGetter<HttpServletRequest> getter =
    new TextMapGetter<HttpServletRequest>() {
      @Override
      public String get(HttpServletRequest carrier, String key) {
        Enumeration<String> headerNames = carrier.getHeaderNames();
        if (headerNames != null) {
          while (headerNames.hasMoreElements()) {
            String headerName = headerNames.nextElement();
            System.out.println("headerNames.nextElement(): " + headerName);
            if (headerName.equals(key)) {
              String headerValue = request.getHeader(headerName);
              System.out.println("headerValue): " + headerValue);
              return headerValue;
            }
          }
        }

        return null;
      }

      @Override
      public Iterable<String> keys(HttpServletRequest carrier) {
        Set<String> set = new HashSet<String>();
        Enumeration<String> headerNames = carrier.getHeaderNames();
        if (headerNames != null) {
          while (headerNames.hasMoreElements()) {
            set.add(headerNames.nextElement());
          }
        }
        return set;
      }
    };
Context extractedContext =
    GlobalOpenTelemetry.getPropagators()
        .getTextMapPropagator()
        .extract(Context.current(), request, getter);
}`

At server, i am not able to get parent span. Kindly help on this.

**Comment by jkwatson** [comment]
How have you configured your OpenTelemetry instances on client and server? Have you looked at the examples in this repository?

**Reply by mayur-more** [reply]
Yes I had a look at examples given under examples/distro, but didn't help.
About OpenTelemetry configuration, I am not sure about this. I am using  GlobalOpenTelemetry.

**Reply by jkwatson** [reply]
If you haven't configured the GlobalOpenTelemetry instance, it will be a no-op and hence show the behavior you describe. 

Take another look at the examples and look how they are configuring an SDK.

**Reply by mayur-more** [reply]
can you please guide me to point specific example.

**Reply by jkwatson** [reply]
I think this does precisely what you are attempting:  https://github.com/open-telemetry/opentelemetry-java/tree/main/examples/http

**Reply by mayur-more** [reply]
In my auto-instrumentation example, I have used the code in given ExampleConfiguration.java, But now i am not able to get my parent spans at both end(client and server)

**Reply by jkwatson** [reply]
wait... if you're using the auto-instrumentation agent, that's a completely different story, and our examples do not make sense at all. 

I think this conversation should be restarted over in the opentelemetry-java-instrumentation repository, because I had been assuming you were just using the bare API/SDK.

**Reply by jkwatson** [reply]
If you're using the auto-instrumentation agent, you shouldn't need to do anything at all but just start it up. You don't need to do any extra work to make propagation work.

**Reply by mayur-more** [reply]
I did it in first way but i didn't get my span in one trace. I am getting it in 2 different traces. My  requirement is, these 2 traces should combine, as it is covering one operation  .

**Reply by jkwatson** [reply]
Again, if you're using auto-instrumentation, it should all just work. You shouldn't have to write a single line of code to do anything, and if you do, it will probably backfire and result in broken traces, like you describe.

**Reply by jkwatson** [reply]
If you can post a full example of what isn't working correctly (if it still isn't working after removing all the code that you posted above), then I'm sure the instrumentation folks would be happy to take a look.

**Reply by mayur-more** [reply]
Following is my code for client and server services for auto-instrumentation:
I have added two modules(client and server) under opentelemetry-java-instrumentation/instrumentation/ 
This is code before i have added code for context propagation.
Client Advice class:
  public static class ClientAdvice {
    @Advice.OnMethodEnter
    public static void onEnter(
        @Advice.Origin String origin,
        @Advice.Local("otelContext") Context context,
        @Advice.Local("otelScope") Scope scope) {
      
      String subStr = origin.substring(0, origin.indexOf("(") + 1);
      String operation = subStr.substring(subStr.lastIndexOf(".") + 1, subStr.indexOf("("));
      
      OpenTelemetry openTelemetry = SdkConfiguration.initOpenTelemetry();
      Tracer tracer = openTelemetry.getTracer("test");
      Span span = tracer.spanBuilder("/" + operation).setSpanKind(SpanKind.CLIENT).startSpan();
      scope = span.makeCurrent();

      if (span != null) {
        span.setAttribute("method_name", origin);
      }
    }

    @Advice.OnMethodExit
    public static void onExit(
        @Advice.Origin String origin,
        @Advice.Local("otelContext") Context context,
        @Advice.Local("otelScope") Scope scope) {
      
      Span span = Span.current();
      scope = span.makeCurrent();
      scope.close();
      span.end();
    }
  }

Server Advice class:
  public static class CidsAdvice {
    @Advice.OnMethodEnter
    public static void onEnter(@Advice.Origin String origin) {
      
      String subStr = origin.substring(0, origin.indexOf("(") + 1);
      String operation = subStr.substring(subStr.lastIndexOf(".") + 1, subStr.indexOf("("));

      Tracer tracer = GlobalOpenTelemetry.getTracer("test");
      Span span = tracer.spanBuilder("/" + operation).startSpan();
      Scope scope = span.makeCurrent();

      if (span != null) {
        span.setAttribute("method_name", origin);
      }
    }

    @Advice.OnMethodExit
    public static void onExit(@Advice.Origin String origin) {
      
      Span span = Span.current();
      Scope scope = span.makeCurrent();
      scope.close();
      span.end();
    }
  }

**Reply by jkwatson** [reply]
If you're in the agent, you must not configure your own copy of the SDK. If you need assistance writing agent instrumentation please open a discussion in the open-telemetry/opentelemetry-java-instrumentation repository.

**Reply by mayur-more** [reply]
Thanks. I have created one: https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/3128

**Reply by hkdev27** [reply]
Context injector class:
@Override
	public UserBillingAddress getUserBillingAddress(long id) {
		
		  Span child = tracer.spanBuilder("user-billing-address-service").setSpanKind(SpanKind.CLIENT).startSpan();
		  child.addEvent("User billing address with id-service",Instant.EPOCH);
		  child.setAttribute(TRACE_ID,child.getSpanContext().getTraceId());
		  MDC.put(TRACE_ID,child.getSpanContext().getTraceId());
		 
		log.info("User billing address with id: {}", id);
		try(Scope scope = child.makeCurrent()) {
			
			TextMapSetter<HttpURLConnection> setter =
					  new TextMapSetter<HttpURLConnection>() {
					    @Override
					    public void set(HttpURLConnection carrier, String key, String value) {
					        carrier.setRequestProperty(key, value);
					    }
					};
					URL url = new URL("http://localhost:8081/address");
			TextMapPropagator textMapPropagator = W3CTraceContextPropagator.getInstance();
			HttpURLConnection  httpURLConnection = (HttpURLConnection) url.openConnection();
		     textMapPropagator.inject(Context.current(), httpURLConnection, setter);
			Optional<UserEntity> entity = userRepository.findById(id);
			if(entity.isPresent()) {
				Address address = addressServiceProxy.findById(id);
				User user = convertToDto(entity.get());
				return new UserBillingAddress(user.getId(), user.getName(), address);
			}
			  }catch(Exception exception) { 
				  log.error("Address Service is down"); 
				  }finally
			  { child.end(); }
			 
		return null;
	}

Context extractor class:

TextMapPropagator textMapPropagator = W3CTraceContextPropagator.getInstance();
        
		TextMapGetter<HttpServletRequest> getter =
				  new TextMapGetter<HttpServletRequest>() {
				    @Override
				    public String get(HttpServletRequest carrier, String key) {
				      if (StringUtils.isNoneEmpty(carrier.getHeader(key))) {
				        return carrier.getHeader(key);
				      }
				      return null;
				    }

					@Override
					public Iterable<String> keys(HttpServletRequest carrier) {
						return null;
					}
				};

		   // Extract and store the propagated span's SpanContext and other available concerns
		   // in the specified Context.
		   Context context = textMapPropagator.extract(Context.current(), request, getter);

I'm getting null for request, wherever I am seeing extract is used as follows but for me HttpExchange is not visible( compilation error)

TextMapGetter<HttpExchange> getter =
  new TextMapGetter<>() {
    @Override
    public String get(HttpExchange carrier, String key) {
      if (carrier.getRequestHeaders().containsKey(key)) {
        return carrier.getRequestHeaders().get(key).get(0);
      }
      return null;
    }

   @Override
   public Iterable<String> keys(HttpExchange carrier) {
     return carrier.getRequestHeaders().keySet();
   }
};

Can you help me out.

**Reply by jkwatson** [reply]
It looks like your code is expecting an HttpServletRequest carrier, not an HttpExchange. Make sure you're using the same framework in both parts of your code.
